### PR TITLE
Adding FNO and DeepONets to Neuromancer

### PR DESCRIPTION
**Pull request by Parv621**
This PR updates **#263**.

It introduces **DeepONets** and **Fourier Neural Operators (FNOs)** into **Neuromancer**, with integrations to the external libraries **[neuralop](https://github.com/neuraloperator/neuraloperator)** and **[DeepXDE](https://deepxde.readthedocs.io/en/latest/)**

Integration is provided via lightweight wrappers that preserve Neuromancer’s abstractions while maintaining compatibility with upstream libraries.

---

### Operator Integrations

1. **FNO Wrapper**
   `src/neuromancer/modules/operators.py`
   A thin wrapper enabling direct use of `neuralop` models inside Neuromancer. The wrapper attaches the required `_metadata` for robust checkpointing across PyTorch versions. The design generalizes to any operator provided by `neuralop`.

2. **Native DeepONet (Cartesian Product)**
   `src/neuromancer/modules/operators.py`
   Native implementation of DeepONet with Cartesian-product evaluation, consistent with standard operator-learning formulations.

3. **DeepXDE Wrapper**
   `src/neuromancer/modules/operators.py`
   Wrapper for DeepONet-style models implemented in DeepXDE. Supports both
   `dde.nn.DeepONetCartesianProd` and `dde.nn.DeepONet`.

4. **Custom Operator Losses**
   `src/neuromancer/modules/operators.py`
   Implementation of `LpLoss` and `H1Loss`, suitable for grid-based operator learning.

---

### Training and Validation Infrastructure

5. **Validation Dataset Support During Training**
   Changes in `trainer.py`, `problem.py`, and `callbacks.py` to support validation datasets during training (inspired by #168).
   A new `LossHistoryCallback` is added to `callbacks.py` to track and plot training and validation loss per epoch.

---

## Tests Added

The following unit tests were added to validate new functionality:

- **`test_trainer_optional_dev_data`** (`test_trainer.py`)  
  Verifies that the training loop runs correctly both with and without a validation (dev) dataset.

- **`test_problem_predict`** (`test_problem.py`)  
  Tests the `Problem.predict` method, ensuring gradients are disabled during inference and that input/output keys are handled correctly.

- **`test_strip_state` and `test_strip_load`** (`test_operators.py`)  
  Validate the `_StripMetadataMixin` used by the `neuralop` wrapper to ensure metadata is correctly stripped during save and load operations.

- Tests were added for `DeepXDEWrapper` and `DeepONetCartesianProd` in `operators.py`, (`test_operators.py`)  .

---

### DeepONet Examples

1. **Antiderivative (Aligned GRF Dataset)**
   `examples/neural_operators/Part_1_DeepONet_antiderivative_aligned.ipynb`
   Inspiration: DeepXDE aligned antiderivative demo and Lu et al. (2021).

2. **Antiderivative (Unaligned GRF Dataset)**
   `examples/neural_operators/Part_2_DeepONet_antiderivative_unaligned.ipynb`
   Inspiration: DeepXDE unaligned antiderivative demo.

3. **1D Advection equation (Aligned initial conditions from GRF fields)**
   `examples/neural_operators/Part_6_PIDeepONet_Advection_aligned.ipynb`
   Inspiration: DeepXDE aligned 1D Advection demo. 

---

### FNO Examples

1. **1D Allen–Cahn Equation (from-scratch FNO)**
   `examples/neural_operators/Part_4_FNO_1DAllenCahn.ipynb`
   Inspiration: ETH “Operator Learning – FNO” tutorial and the original FNO paper.

2. **Darcy Flow (FNO, Neuromancer pipeline)**
   `examples/neural_operators/Part_5_FNO_DarcyFlow.ipynb`
   Inspiration: `neuralop` Darcy Flow example and NVIDIA’s FNO Darcy guide.

3. **Diffusion Equation (FNO, Neuromancer pipeline)**
   `examples/neural_operators/Part_6_FNO_DiffusionEquation.ipynb`
   Inspiration: `examples/PDEs/Part_1_PINN_DiffusionEquation.ipynb`.

4. **Shallow Water equations using Spherical FNO (SFNO, Neuromancer pipeline)**
   `examples/neural_operators/Part_7_SFNO_ShallowWater.ipynb`
   Inspiration: `neuralop` SFNO on the spherical Shallow Water equations.

---

## Notebooks I could not implement (DeepONet)

The following notebooks are not included as they do not achieve accurate solutions using standard MLP DeepONets:

* **Cartesian-Product DeepONet – Darcy Flow**
* **Cartesian-Product DeepONet – Diffusion Equation**
* **Vanilla DeepONet – Darcy Flow**
* **Vanilla DeepONet – Diffusion Equation**

These are moved to another SOLARIS branch (`feature/DeepONet_notebooks_debug`); we could revisit them later.

Some papers try to solve DarcyFlow using NOs, might be useful for debugging. [NVIDIA example on Darcy Flow using DeepONet](https://docs.nvidia.com/physicsnemo/25.08/physicsnemo-sym/user_guide/neural_operators/deeponet.html#problem-3-darcy-flow-data-informed) uses a convnet-based model [pix2pix](https://docs.nvidia.com/physicsnemo/25.08/physicsnemo-sym/user_guide/theory/architectures.html#pix2pix-net) in the branch net. Another paper (https://www.mdpi.com/2227-9717/13/9/2754) solves DarcyFlow using various methods

### CODE REVIEW

**Review by drgona — COMMENTED** [review]
we can add the `LossHistoryCallback` class to callbacks.py

Since users will start from the first example, it would be good if the Part 1 example has extensive method markdown, similar to Part 3.

**Comment by Parv621** [comment]
> we can add the `LossHistoryCallback` class to callbacks.py
> 
> Since users will start from the first example, it would be good if the Part 1 example has extensive method markdown, similar to Part 3.

Added in callback in bf02c6119214968d731275a52e1dd1c166962126, updated Part 1 in 002d89c1b5b237302dbeb3ac55da44d2e31d16e3

File: pyproject.toml [hunk 1]
```diff
@@ -43,14 +43,14 @@ dependencies = [
     "six",
     "pydot==1.4.2",
     "pyts",
-    "torch",
+    "torch>=1.8",
```
Comment (Parv621) [inline pyproject.toml:46]: For ifft in PyTorch needed for FNOs

File: src/neuromancer/modules/blocks.py
Comment (Parv621) [inline src/neuromancer/modules/blocks.py:1]: Formatting changes only (linter)

File: pyproject.toml (hunk 1 above)
Comment (Parv621) [reply pyproject.toml:46]: Should I add dependencies for neuralop and deepxde to the environment? Or would you like to keep it flexible?

Comment (drgona) [reply pyproject.toml:46]: It depends how many and what dependencies there are. If it is a lot, we can add the necessary dependencies for neuralop and deepxde to examples dependencies in our toml file.

https://github.com/pnnl/neuromancer/blob/master/pyproject.toml#L69

**Comment by drgona** [comment]
lets run all the unit tests on your on this PR to make sure we are not braking the core

**Comment by drgona** [comment]
Well done with the examples!
Few minor comments on notebooks below.

1. Lets make sure the examples are ordered withour overlapping Part numbers. Lets consolidate examples to make them more compact. 
2. Each new example should add markdown to new features or architectures it introduces compared to previous examples.
3. Can we have antiderivative example to introduce simple FNO as well?

**Part_1_DeepONet_antiderivative_aligned.ipynb**
markdown cell with Operator evaluation: lets specify `b` for grid point evaluations, similarly we define `m` for sensor points
lets explain what is the attribute  "Glorot normal" in DeepXDE DeepOnet

**Part_2_DeepONet_antiderivative_unaligned.ipynb**
it would be nice to have base DeepONet class in neuromancer and compare with DeepXDE in this example similarly as in part 1. Nice to have if easy to implement, but not critical.

**Part_5_FNO_DarcyFlow.ipynb**
If we don't introduce FNO on antiderivative, we can use this example as main intro to FNO and make it part 3 and potentially merge with **Part_3_DeepONet_DarcyFlow.ipynb** to compare FNO vs DeepOnet

**Part_3_DeepONet_DarcyFlow.ipynb**
this notebook introduces for the first time `DeepONetBatchWrapper ` but it is not explained, what is the difference between the `DeepXDEWrapper`?

It is not apparent from the notebooks what is the main difference between:
**Part_3_DeepONetCartesianProd_DarcyFlow.ipynb**
**Part_3_DeepONet_DarcyFlow.ipynb**
Maybe we can merge these two into one notebook and compare cartesiaon prod with non-catesian DeepOnet and focus on explaining the main differences and their implications.
Currently both notebooks have large cells of code without markdown.

Similarly:
**Part_6_FNO_DiffusionEquation.ipynb**
**Part_3_DeepONet_DiffusionEquation.ipynb**
**Part_3_DeepONetCartesianProd_DiffusionEquation.ipynb**
These can be merged into single exmaple that compares FNO vs DeepONet on diffusion example and focus on explaining main differences.

**Part_3_DeepONetCartesianProd_DiffusionEquation.ipynb** this title does not match the content which shows FNO. It looks like deprecated file to be deleted.

**Part_4_FNO_1DAllenCahn.ipynb**
The role of convolution layers should be explained.

File: src/neuromancer/callbacks.py
Comment (drgona) [inline src/neuromancer/callbacks.py:1]: lets make sure this dependency is included in our toml file:
from pathlib import Path

File: src/neuromancer/problem.py
Comment (drgona) [inline src/neuromancer/problem.py:1]: since you introduce change to the core abstraction, we need to run all the tests to make sure we are not breaking the core

File: src/neuromancer/trainer.py
Comment (drgona) [inline src/neuromancer/trainer.py:1]: similarly here, lets run all the tests to make sure these changes to the core abstraction are backward compatible

**Comment by drgona** [comment]
> Well done with the examples! Few minor comments on notebooks below.
> 
> 1. Lets make sure the examples are ordered withour overlapping Part numbers. Lets consolidate examples to make them more compact.
> 2. Each new example should add markdown to new features or architectures it introduces compared to previous examples.
> 3. Can we have antiderivative example to introduce simple FNO as well?
> 
> **Part_1_DeepONet_antiderivative_aligned.ipynb** markdown cell with Operator evaluation: lets specify `b` for grid point evaluations, similarly we define `m` for sensor points lets explain what is the attribute "Glorot normal" in DeepXDE DeepOnet
> 
> **Part_2_DeepONet_antiderivative_unaligned.ipynb** it would be nice to have base DeepONet class in neuromancer and compare with DeepXDE in this example similarly as in part 1. Nice to have if easy to implement, but not critical.
> 
> **Part_5_FNO_DarcyFlow.ipynb** If we don't introduce FNO on antiderivative, we can use this example as main intro to FNO and make it part 3 and potentially merge with **Part_3_DeepONet_DarcyFlow.ipynb** to compare FNO vs DeepOnet
> 
> **Part_3_DeepONet_DarcyFlow.ipynb** this notebook introduces for the first time `DeepONetBatchWrapper ` but it is not explained, what is the difference between the `DeepXDEWrapper`?
> 
> It is not apparent from the notebooks what is the main difference between: **Part_3_DeepONetCartesianProd_DarcyFlow.ipynb** **Part_3_DeepONet_DarcyFlow.ipynb** Maybe we can merge these two into one notebook and compare cartesiaon prod with non-catesian DeepOnet and focus on explaining the main differences and their implications. Currently both notebooks have large cells of code without markdown.
> 
> Similarly: **Part_6_FNO_DiffusionEquation.ipynb** **Part_3_DeepONet_DiffusionEquation.ipynb** **Part_3_DeepONetCartesianProd_DiffusionEquation.ipynb** These can be merged into single exmaple that compares FNO vs DeepONet on diffusion example and focus on explaining main differences.
> 
> **Part_3_DeepONetCartesianProd_DiffusionEquation.ipynb** this title does not match the content which shows FNO. It looks like deprecated file to be deleted.
> 
> **Part_4_FNO_1DAllenCahn.ipynb** The role of convolution layers should be explained.



FNO examples:
**Part_4_FNO_1DAllenCahn.ipynb** -> **Part_3_FNO_1DAllenCahn.ipynb**
**Part_4_FNO_DarcyFlow.ipynb** ->  **Part_4_FNO_DarcyFlow.ipynb** - explain the embeddings
**Part_6_FNO_DiffusionEquation.ipynb** -> **Part_5_FNO_DiffusionEquation.ipynb** explain the diff with **Part_4_FNO_DarcyFlow.ipynb**

File: pyproject.toml (hunk 1 above)
Comment (Parv621) [reply pyproject.toml:46]: So, the `deepxde` library can be added like that, as we use it only in examples.
The `neuraloperator` library is being used here https://github.com/SOLARIS-JHU/neuromancer_NO/blob/feature/FNO/src/neuromancer/modules/operators.py#L6, so it will either need to be added as a main dependency. Or we could do something like this in pyproject.toml file, loading this only for operators.

```toml
examples = [
    "deepxde",
    "imageio",
    "neuraloperator",
]
[project.optional-dependencies]
operators = [
  "neuraloperator",
]
```

Let me know what you prefer!

File: src/neuromancer/callbacks.py
Comment (Parv621) [reply src/neuromancer/callbacks.py:1]: pathlib is a standard library module. No dependency entry is required.

File: src/neuromancer/problem.py
Comment (Parv621) [reply src/neuromancer/problem.py:1]: image

File: src/neuromancer/trainer.py
Comment (Parv621) [reply src/neuromancer/trainer.py:1]: > similarly here, lets run all the tests to make sure these changes to the core abstraction are backward compatible

this passed with a few warnings

**Comment by Parv621** [comment]
Updated the PR with SFNO examples as well ([SFNO on the spherical Shallow Water equations](https://neuraloperator.github.io/dev/auto_examples/models/plot_SFNO_swe.html))

**Review by drgona — CHANGES_REQUESTED** [review]
**Examples**
[Part_2_DeepONet_antiderivative_unaligned](https://github.com/SOLARIS-JHU/neuromancer_NO/blob/feature/FNO/examples/neural_operators/Part_2_DeepONet_antiderivative_unaligned.ipynb)
typos in markdown: 
"We use dde.nn.DeepONetCartesianProd to create the DeepONet directly for DeepXDE library" -> this should be 'dde.nn.DeepONet'

[Part_3_FNO_1DAllenCahn](https://github.com/SOLARIS-JHU/neuromancer_NO/blob/feature/FNO/examples/neural_operators/Part_3_FNO_1DAllenCahn.ipynb)

typo in title: Operator Learing -> Operator Learning

typos in markdown: Problem Description - Allen-Cahn equation
- "$(u^3 - u)$ is the derivative of the double-well potential..." should be it own bullet point 
- also lets define the diffusion term explicitly 
- In Allen–Cahn the coefficient \epsilon^2 is typically multiplying the diffusion term, alternatively we can write 1/\epsilon^2 in front of the potential

specify the data: domain, boundary conditions, what input and output represent

[Part_4_FNO_DarcyFlow](https://github.com/SOLARIS-JHU/neuromancer_NO/blob/feature/FNO/examples/neural_operators/Part_4_FNO_DarcyFlow.ipynb)

- The problem description can be confusing. Your Problem Description introduces “1D Darcy law” and then immediately uses the 2D elliptic PDE form. So it is confusing whether the example is 1D or 2D. We need to clarify.
- specify the data: domain, boundary conditions, what input and output represent

[Part_5_FNO_DiffusionEquation](https://github.com/SOLARIS-JHU/neuromancer_NO/blob/feature/FNO/examples/neural_operators/Part_5_FNO_DiffusionEquation.ipynb)

- add relative path hyperlink to Darcy Flow example
- in the introduction markdown you use u, and in the loss definition you use y for outputs. Consolidate, double check the same in other examples

[Part_6_PIDeepONet_Advection_aligned](https://github.com/SOLARIS-JHU/neuromancer_NO/blob/feature/FNO/examples/neural_operators/Part_6_PIDeepONet_Advection_aligned.ipynb)

- This equation could benefit from mass conservation loss. 
$\displaystyle \int u(x,t)\,dx = \text{const.}$
this could be implemented by: 
1. constraining mass to equal the initial mass
2. penalize time derivative of mass $\displaystyle \int_\Omega \partial_t u(x,t)\,dx \approx 0$
3. Normalize mass via the relative mass conservation loss as
\[
L
= \frac{1}{N_t}\sum_{j=1}^{N_t}
\left(
\frac{M_\theta(t_j) - M_0}{|M_0| + \varepsilon}
\right)^2.
\]
$\varepsilon > 0$ is a small constant added for numerical stability

- we coul plot phase error: shift error vs time
- Advection without specifying BCs is ill-posed.  Why are we not imposing BC in the loss?

[Part_7_SFNO_ShallowWater](https://github.com/SOLARIS-JHU/neuromancer_NO/blob/feature/FNO/examples/neural_operators/Part_7_SFNO_ShallowWater.ipynb)

- Add didactical markdown to explain the SFNO architectyre
- Add reference to the [paper](https://arxiv.org/abs/2306.03838)
- explain the PDE problem in sufficient detail, add equations similarly as in other examples
- define operator-learning task


**Core**
Core changes in trainer.py, problem.py:
Add 1–2 lightweight unit tests that verify:
- training loop accepts dev_loader/validation dataset 
- callback doesn’t crash when dev data is absent/present
- Ensure default behavior is unchanged when no validation dataset is provided.

**Comment by Parv621** [comment]
Everything apart from changes suggested in [Part_6_PIDeepONet_Advection_aligned](https://github.com/SOLARIS-JHU/neuromancer_NO/blob/feature/FNO/examples/neural_operators/Part_6_PIDeepONet_Advection_aligned.ipynb) is complete. 

Good point regarding the ill-posedness. The BCs are actually handled implicitly via the data distribution. By using the ExpSineSquared kernel, all generated initial conditions are periodic. This effectively imposes Periodic BCs, resolving the ill-posedness.

I will add a small markdown to the notebook

**Comment by drgona** [comment]
> * n could benefit from mass conservation loss.

no need to apply the suggested changes to [Part_6_PIDeepONet_Advection_aligned](https://github.com/SOLARIS-JHU/neuromancer_NO/blob/feature/FNO/examples/neural_operators/Part_6_PIDeepONet_Advection_aligned.ipynb) - small markdown will be nice

lets focus on the pytests so we can deploy this PR today :)

**Comment by Parv621** [comment]
> > * n could benefit from mass conservation loss.
> 
> no need to apply the suggested changes to [Part_6_PIDeepONet_Advection_aligned](https://github.com/SOLARIS-JHU/neuromancer_NO/blob/feature/FNO/examples/neural_operators/Part_6_PIDeepONet_Advection_aligned.ipynb) - small markdown will be nice
> 
> lets focus on the pytests so we can deploy this PR today :)

Added the following pytests:
- `test_trainer_optional_dev_data` in test_trainer.py to check for running with and without dev data
- `test_problem_predict` in test_problem.py to check the `predict` method in problem.py (grad off, check input/ouput keys)
- `test_strip_state` and `test_strip_load` in test_operators.py, to check the methods in the neualoperator library wrapper `class _StripMetadataMixin` in operators.py

I will also update the main description accordingly

**Comment by Parv621** [comment]
Pytests ran:

test_problem.py, test_operators.py
image

test_trainer.py (reletively time consuming)
image

**Review by drgona — APPROVED** [review]
tests passing on my end, PR is ready to bedeployed
